### PR TITLE
[main][feat] xALPACAv2

### PR DESCRIPTION
**Pull request by alpanaca**
## Description
New locking contract for xALPACA. 

## Why?
Refer to AIP-24

## ChangeLogs:
- xALPACAv2 contract
- [wip] pending for miniFL integration

### CODE REVIEW

File: contracts/8.19/xALPACAv2.sol [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+    // effect
+    unchecked {
+      userLockAmounts[msg.sender] -= _amount;
+      totalLocked -= _amount;
+    }
+
+    UnlockRequest memory _request = UnlockRequest({
+      amount: _amount,
+      unlockTimestamp: uint64(block.timestamp + delayUnlockTime),
```
[the remark this answers is not in the document]
Comment (alpanaca) [reply contracts/8.19/xALPACAv2.sol:124]: Don't think we need to if we cap the delayUnlockTime. Shouldn't overflow the uint64

File: contracts/8.19/xALPACAv2.sol [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    token.safeTransfer(_to, _amount);
+
+    emit LogWithdrawReserve(_to, _amount);
+  }
+
+  /// @notice Owner set early withdraw fee bps per sec
+  /// @param _newFeePerPerDay The new early withdrawal bps per day
+  function setEarlyWithdrawFeeBpsPerDay(uint256 _newFeePerPerDay) external onlyOwner {
+    emit LogSetEarlyWithdrawFeeBpsPerDay(earlyWithdrawFeeBpsPerDay, _newFeePerPerDay);
+    earlyWithdrawFeeBpsPerDay = _newFeePerPerDay;
```
[the remark this answers is not in the document]
Comment (alpanaca) [reply contracts/8.19/xALPACAv2.sol:257]: yeah. should have max

File: contracts/8.19/xALPACAv2.sol [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    if (request.status != 0) {
+      revert xALPACAv2_InvalidStatus();
+    }
+
+    if (block.timestamp >= request.unlockTimestamp) {
+      revert xALPACAv2_UnlockTimeReached();
+    }
+    // effect
+    // fee = (amount * fee per day * second until unlock) / second in day
+    uint256 _earlyWithdrawalFee = (
```
[the remark this answers is not in the document]
Comment (alpanaca) [reply contracts/8.19/xALPACAv2.sol:185]: No need if we cap the fee per day

File: contracts/8.19/xALPACAv2.sol [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+      (request.amount * earlyWithdrawFeeBpsPerDay * (request.unlockTimestamp - block.timestamp))
+    ) / (10000 * 1 days);
+    uint256 _amountToUser = request.amount - _earlyWithdrawalFee;
+
+    request.status = 1; // withdrawn
+    feeReserve += _earlyWithdrawalFee;
```
Comment (alpanaca) [inline contracts/8.19/xALPACAv2.sol:191]: On a separate note, do you think it's better to just transfer out here? to reduce our operation.

@jr-alpaca

File: contracts/8.19/xALPACAv2.sol [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+  // Total amount of token locked
+  uint256 public totalLocked;
+
+  // Flag to allow emergency withdraw
+  uint256 public breaker;
+
+  // Protocol Early Withdrawal Fee treasury
+  address public feeTreasury;
+
+  // penalty per sec
```
Comment (jr-alpaca) [inline contracts/8.19/xALPACAv2.sol:80]:
```suggestion
  // penalty per day
```

**Review by jr-alpaca — APPROVED** [review]

**Review by wow-alpaca — APPROVED** [review]

**Review by jr-alpaca — APPROVED** [review]

**Pull request merged**